x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java

```diff
@@ -427,4 +434,43 @@ public static boolean isPushableFieldAttribute(Expression exp, Predicate<FieldAt
         return false;
     }
 
+    private static class SpatialFromDocValuesExtraction extends OptimizerRule<AggregateExec> {
+        @Override
+        protected PhysicalPlan rule(AggregateExec aggregate) {
+            var foundAttribute = new Holder<Attribute>(null);
+
+            PhysicalPlan plan = aggregate.transformDown(UnaryExec.class, exec -> {
```

Comment:
This is the heart of the solution for figuring out when to use doc-values and when to use source values. It only runs on the local node, so we only consider doc-values on the local node, and it re-writes spatial aggregations to mark fields as doc-values so the right spatial aggregation functions are executed, as well as finding the fields involved, and re-writing the FieldExtractExec to load that field as doc-values from the index.